### Load tables: prep_orders/prep_order_items/prep_seller

In [302]:
from sqlalchemy import create_engine
from sqlalchemy import text 
import pandas as pd

In [303]:
from dotenv import dotenv_values

config = dotenv_values()

pg_user = config['POSTGRES_USER']  
pg_host = config['POSTGRES_HOST']
pg_port = config['POSTGRES_PORT']
pg_db = config['POSTGRES_DB']
pg_schema = config['POSTGRES_SCHEMA']
pg_pass = config['POSTGRES_PASS']

In [304]:
url = f'postgresql://{pg_user}:{pg_pass}@{pg_host}:{pg_port}/{pg_db}' #the same like version 1

engine = create_engine(url, echo=False) 

my_schema = 'team_jjat'

with engine.begin() as conn: 
    result = conn.execute(text(f'SET search_path TO {my_schema};'))

In [305]:
prep_seller = pd.read_sql('SELECT * FROM prep_seller;', con=engine)
prep_seller.head()

seller_id seller_zip_code_prefix  seller_city  \
0  00065220becb8785e2cf78355eb9bf68                   None         None   
1  0015a82c2db000af6aaaf3ae2ecb0532                  09080  santo andre   
2  001cca7ae9ae17fb1caed9dfb1094831                  29156    cariacica   
3  001e6ad469a905060d959994f1b41e4f                  24754  sao goncalo   
4  002100f778ceb8431b7a1020ff7ab48f                  14405       franca   

  seller_state            won_date business_segment_old      lead_type  \
0         None 2018-07-10 11:34:50          handcrafted  online_medium   
1           SP                 NaT                 None           None   
2           ES                 NaT                 None           None   
3           RJ                 NaT                 None           None   
4           SP                 NaT                 None           None   

   has_company  has_gtin business_type business_segment  \
0        False     False      reseller    Home & Living   
1         True      True          None             None   
2         True      True          None             None   
3         True      True          None             None   
4         True      True          None             None   

      business_segment_pred business_segment_incl_pred  
0                      None              Home & Living  
1  Electronics & Technology   Electronics & Technology  
2        Construction Tools         Construction Tools  
3                     Sport                      Sport  
4             Home & Living              Home & Living

In [306]:
prep_order_items = pd.read_sql('SELECT * FROM prep_order_items;', con=engine)
prep_order_items.head()

order_id                        product_id  \
0  00010242fe8c5a6d1ba2dd792cb16214  4244733e06e7ecb4970a6e2683c13e61   
1  00018f77f2f0320c557190d7a144bdd3  e5f2d52b802189ee658865ca93d83a8f   
2  000229ec398224ef6ca0657da4fc703e  c777355d18b72b67abbeef9df44fd0fd   
3  00024acbcdf0a6daa1e931b038114c75  7634da152a4610f1595efa32f14722fc   
4  00042b26cf59d7ce69dfabb4e55b4fd9  ac6c3623068f30de03045865e4e10089   

   order_item_id                         seller_id  shipping_limit_date  \
0              1  48436dade18ac8b2bce089ec2a041202  2017-09-19 09:45:35   
1              1  dd7ddc04e1b6c2c614352b383efe2d36  2017-05-03 11:05:13   
2              1  5b51032eddd242adc84c38acab88f23d  2018-01-18 14:48:30   
3              1  9d7a1d34a5052409006425275ba1c2b4  2018-08-15 10:10:18   
4              1  df560393f3a51e74553ab94004ba5c87  2017-02-13 13:57:51   

    price  freight_value  
0   58.90          13.29  
1  239.90          19.93  
2  199.00          17.87  
3   12.99          12.79  
4  199.90          18.14

In [307]:
prep_orders = pd.read_sql('SELECT * FROM prep_orders;', con=engine)
prep_orders.head()

order_id                       customer_id  \
0  2e7a8482f6fb09756ca50c10d7bfc047  08c5351a6aca1c1589a38f244edeee9d   
1  e5fa5a7210941f7d56d0208e4e071d35  683c54fc24d40ee9f8a6fc179fd9856c   
2  809a282bbd5dbcabb6f2f724fca862ec  622e13439d6b5a0b486c435618b2679e   
3  bfbd0f9bdef84302105ad712db648a6c  86dc2ffce2dfff336de2f386a786e574   
4  71303d7e93b399f5bcd537d124c0bcfa  b106b360fe2ef8849fbbd056f777b4d5   

  order_status order_purchase_timestamp   order_approved_at  \
0      shipped      2016-09-04 21:15:19 2016-10-07 13:18:03   
1     canceled      2016-09-05 00:15:34 2016-10-07 13:17:15   
2     canceled      2016-09-13 15:24:19 2016-10-07 13:16:46   
3    delivered      2016-09-15 12:16:38 2016-09-15 12:16:38   
4     canceled      2016-10-02 22:07:52 2016-10-06 15:50:56   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2016-10-18 13:14:51                           NaT   
1                          NaT                           NaT   
2                          NaT                           NaT   
3          2016-11-07 17:11:53           2016-11-09 07:47:38   
4                          NaT                           NaT   

  order_estimated_delivery_date  
0                    2016-10-20  
1                    2016-10-28  
2                    2016-09-30  
3                    2016-10-04  
4                    2016-10-25

### Inspect data

In [308]:
prep_seller.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3557 entries, 0 to 3556
Data columns (total 13 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   seller_id                   3557 non-null   object        
 1   seller_zip_code_prefix      3095 non-null   object        
 2   seller_city                 3095 non-null   object        
 3   seller_state                3095 non-null   object        
 4   won_date                    842 non-null    datetime64[ns]
 5   business_segment_old        842 non-null    object        
 6   lead_type                   842 non-null    object        
 7   has_company                 3557 non-null   bool          
 8   has_gtin                    3557 non-null   bool          
 9   business_type               842 non-null    object        
 10  business_segment            842 non-null    object        
 11  business_segment_pred       2656 non-null   object      

#### order id in the order item table is duplicated, one order id can have different products from different sellers.

In [309]:
prep_order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102425 entries, 0 to 102424
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             102425 non-null  object 
 1   product_id           102425 non-null  object 
 2   order_item_id        102425 non-null  int64  
 3   seller_id            102425 non-null  object 
 4   shipping_limit_date  102425 non-null  object 
 5   price                102425 non-null  float64
 6   freight_value        102425 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 5.5+ MB


In [310]:
unique_orders = prep_order_items['order_id'].nunique()
unique_orders

98666

In [311]:
total_rows = len(prep_order_items)

In [312]:
percentage_unique_orders = (unique_orders / total_rows) * 100
percentage_unique_orders

96.32999755918965

In [313]:
prep_order_items['order_id'].value_counts()

order_id
ca3625898fbd48669d50701aba51cd5f    8
7d8f5bfd5aff648220374a2df62e84d5    7
77df84f9195be22a4e9cb72ca9e8b4c2    7
ad850e69fce9a512ada84086651a2e7d    7
5efc0b7fe9df7f0c567404abaa4d25fc    6
                                   ..
56d0a27ec26d1644336aff1ec939c089    1
56d01e37808556640e75a40c8202b0a8    1
56ce64b53ec3bb5b0525f21506619f8e    1
56ce3873eb6d633a47581ec221f50a65    1
fffe41c64501cc87c801fd61db3f6244    1
Name: count, Length: 98666, dtype: int64

In [314]:
prep_order_items[prep_order_items['order_id']=='ca3625898fbd48669d50701aba51cd5f']

order_id                        product_id  \
80867  ca3625898fbd48669d50701aba51cd5f  0cf2faf9749f53924cea652a09d8e327   
80868  ca3625898fbd48669d50701aba51cd5f  0de59eddc63167215c972b0d785ffa7b   
80869  ca3625898fbd48669d50701aba51cd5f  1065e0ebef073787a7bf691924c60eeb   
80870  ca3625898fbd48669d50701aba51cd5f  21b524c4c060169fa75ccf08c7da4627   
80871  ca3625898fbd48669d50701aba51cd5f  309dd69eb83cea38c51709d62befe1a4   
80872  ca3625898fbd48669d50701aba51cd5f  4a5c3967bfd3629fe07ef4d0cc8c3818   
80873  ca3625898fbd48669d50701aba51cd5f  5dae498eff2d80057f56122235a36aff   
80874  ca3625898fbd48669d50701aba51cd5f  678c229b41c0e497d35a25a8be1cc631   

       order_item_id                         seller_id  shipping_limit_date  \
80867              2  0b35c634521043bf4b47e21547b99ab5  2018-08-16 02:25:07   
80868              4  0b35c634521043bf4b47e21547b99ab5  2018-08-16 02:25:07   
80869              1  0b35c634521043bf4b47e21547b99ab5  2018-08-16 02:25:07   
80870              8  0b35c634521043bf4b47e21547b99ab5  2018-08-16 02:25:07   
80871             10  0b35c634521043bf4b47e21547b99ab5  2018-08-16 02:25:07   
80872              6  0b35c634521043bf4b47e21547b99ab5  2018-08-16 02:25:07   
80873              5  888faa8bfb0b159c37de6d898b961c31  2018-08-17 02:25:07   
80874              7  888faa8bfb0b159c37de6d898b961c31  2018-08-17 02:25:07   

       price  freight_value  
80867   33.9           1.84  
80868  159.0           3.67  
80869  309.0           1.84  
80870   63.7           0.15  
80871   56.0           3.68  
80872  109.9           0.15  
80873   95.9           0.15  
80874   95.9           0.15

#### order id in the orders table is unique and 99.22% of the order id can be found in the order items table

In [315]:
prep_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  object        
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]
 4   order_approved_at              99281 non-null  datetime64[ns]
 5   order_delivered_carrier_date   97658 non-null  datetime64[ns]
 6   order_delivered_customer_date  96476 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  datetime64[ns]
dtypes: datetime64[ns](5), object(3)
memory usage: 6.1+ MB


In [316]:
# Check if each order in prep_orders exists in prep_order_items
prep_orders['in_order_items'] = prep_orders['order_id'].isin(prep_order_items['order_id'])

In [317]:
# Number of orders that exist in order items
in_items_count = prep_orders['in_order_items'].sum()

# Number of orders not in order items
not_in_items_count = (~prep_orders['in_order_items']).sum()

print(f"Orders in order items: {in_items_count}")
print(f"Orders not in order items: {not_in_items_count}")


Orders in order items: 98666
Orders not in order items: 775


In [318]:
total_orders = len(prep_orders)
percentage_in_items = (in_items_count / total_orders) * 100
percentage_not_in_items = (not_in_items_count / total_orders) * 100

print(f"Percentage in order items: {percentage_in_items:.2f}%")
print(f"Percentage not in order items: {percentage_not_in_items:.2f}%")


Percentage in order items: 99.22%
Percentage not in order items: 0.78%


#### all order id from the order items table are included in the orders table

In [319]:
# Check if each order in prep_orders_items exists in prep_orders
prep_order_items['in_orders'] = prep_order_items['order_id'].isin(prep_orders['order_id'])

In [320]:
# Number of order items whose order_id exists in prep_orders
in_orders_count = prep_order_items['in_orders'].sum()

# Number of order items whose order_id does NOT exist in prep_orders
not_in_orders_count = (~prep_order_items['in_orders']).sum()

print(f"Order items linked to existing orders: {in_orders_count}")
print(f"Order items NOT linked to any order: {not_in_orders_count}")

Order items linked to existing orders: 102425
Order items NOT linked to any order: 0


In [321]:
total_order_items = len(prep_order_items)

percentage_in_orders = (in_orders_count / total_order_items) * 100
percentage_not_in_orders = (not_in_orders_count / total_order_items) * 100

print(f"Percentage in orders: {percentage_in_orders:.2f}%")
print(f"Percentage not in orders: {percentage_not_in_orders:.2f}%")

Percentage in orders: 100.00%
Percentage not in orders: 0.00%


### Merging & cleaning: prep_seller/prep_order_items/prep_orders

In [322]:
import pandas as pd

prep_orders_and_order_items=pd.merge(prep_orders, prep_order_items, how='left', on='order_id')

prep_orders_and_order_items.head()

order_id                       customer_id  \
0  2e7a8482f6fb09756ca50c10d7bfc047  08c5351a6aca1c1589a38f244edeee9d   
1  2e7a8482f6fb09756ca50c10d7bfc047  08c5351a6aca1c1589a38f244edeee9d   
2  e5fa5a7210941f7d56d0208e4e071d35  683c54fc24d40ee9f8a6fc179fd9856c   
3  809a282bbd5dbcabb6f2f724fca862ec  622e13439d6b5a0b486c435618b2679e   
4  bfbd0f9bdef84302105ad712db648a6c  86dc2ffce2dfff336de2f386a786e574   

  order_status order_purchase_timestamp   order_approved_at  \
0      shipped      2016-09-04 21:15:19 2016-10-07 13:18:03   
1      shipped      2016-09-04 21:15:19 2016-10-07 13:18:03   
2     canceled      2016-09-05 00:15:34 2016-10-07 13:17:15   
3     canceled      2016-09-13 15:24:19 2016-10-07 13:16:46   
4    delivered      2016-09-15 12:16:38 2016-09-15 12:16:38   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2016-10-18 13:14:51                           NaT   
1          2016-10-18 13:14:51                           NaT   
2                          NaT                           NaT   
3                          NaT                           NaT   
4          2016-11-07 17:11:53           2016-11-09 07:47:38   

  order_estimated_delivery_date  in_order_items  \
0                    2016-10-20            True   
1                    2016-10-20            True   
2                    2016-10-28            True   
3                    2016-09-30           False   
4                    2016-10-04            True   

                         product_id  order_item_id  \
0  c1488892604e4ba5cff5b4eb4d595400            1.0   
1  f293394c72c9b5fafd7023301fc21fc2            2.0   
2  f3c2d01a84c947b078e32bbef0718962            1.0   
3                               NaN            NaN   
4  5a6b04657a4c5ee34285d1e4619a96b4            3.0   

                          seller_id  shipping_limit_date  price  \
0  1554a68530182680ad5c8b042c3ab563  2016-10-26 18:25:19  39.99   
1  1554a68530182680ad5c8b042c3ab563  2016-10-26 18:25:19  32.90   
2  a425f92c199eb576938df686728acd20  2016-09-19 00:15:34  59.50   
3                               NaN                  NaN    NaN   
4  ecccfa2bb93b34a3bf033cc5d1dcdc69  2016-09-19 23:11:33  44.99   

   freight_value in_orders  
0          31.67      True  
1          31.67      True  
2          15.56      True  
3            NaN       NaN  
4           2.83      True

In [323]:
prep_orders_and_order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103200 entries, 0 to 103199
Data columns (total 16 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       103200 non-null  object        
 1   customer_id                    103200 non-null  object        
 2   order_status                   103200 non-null  object        
 3   order_purchase_timestamp       103200 non-null  datetime64[ns]
 4   order_approved_at              103040 non-null  datetime64[ns]
 5   order_delivered_carrier_date   101398 non-null  datetime64[ns]
 6   order_delivered_customer_date  100195 non-null  datetime64[ns]
 7   order_estimated_delivery_date  103200 non-null  datetime64[ns]
 8   in_order_items                 103200 non-null  bool          
 9   product_id                     102425 non-null  object        
 10  order_item_id                  102425 non-null  float64       
 11  

In [324]:
prep_orders_and_order_items['shipping_limit_date'] = pd.to_datetime(
    prep_orders_and_order_items['shipping_limit_date']
)

In [325]:
prep_orders_and_order_items['order_id'].duplicated().sum()

np.int64(3759)

In [326]:
prep_orders_and_order_items[prep_orders_and_order_items['order_id'].duplicated()]

order_id                       customer_id  \
1       2e7a8482f6fb09756ca50c10d7bfc047  08c5351a6aca1c1589a38f244edeee9d   
41      51725d3e4bdfc97e28b40543310da8a3  a0f95ab56c6d3a5ba0036b7fdfeda07f   
72      8b388d845a3e8700444b971fa2bef902  00474d2582fd72663036795b7ab8cfc1   
79      bcbc0973414b1dd97555e6f8e9713cfa  3f6f000d5b0a959f38481fad4ab7cac2   
80      bcbc0973414b1dd97555e6f8e9713cfa  3f6f000d5b0a959f38481fad4ab7cac2   
...                                  ...                               ...   
102986  4830473362441527781ef6c1860cac4c  72a366446dd3abfc90c598896714d85c   
102990  d7f71d136da2fa678c39380a22d117df  e5a8aaccb766e00177b0a587979cc814   
103033  cb4d0ce750b3542055ae832b4d3bcca4  7a24f850592485f402c7038ca51efe3b   
103063  a63d6f123bbccd71286ef76811cb2a3b  c24bbeb8111fea27eab791a37e665210   
103151  d4fae577806d683110e00e18a5e181be  49a6ae8a95c6a78d90945b983ab1ecfc   

       order_status order_purchase_timestamp   order_approved_at  \
1           shipped      2016-09-04 21:15:19 2016-10-07 13:18:03   
41        delivered      2016-10-04 15:02:53 2016-10-05 17:06:51   
72        delivered      2016-10-04 22:33:44 2016-10-06 03:10:59   
79        delivered      2016-10-04 23:59:01 2016-10-05 03:45:41   
80        delivered      2016-10-04 23:59:01 2016-10-05 03:45:41   
...             ...                      ...                 ...   
102986    delivered      2018-08-26 11:02:58 2018-08-26 11:15:27   
102990    delivered      2018-08-26 11:44:34 2018-08-26 11:55:16   
103033    delivered      2018-08-26 21:01:14 2018-08-26 21:15:10   
103063    delivered      2018-08-27 12:39:11 2018-08-29 03:10:17   
103151    delivered      2018-08-28 19:32:05 2018-08-29 03:15:36   

       order_delivered_carrier_date order_delivered_customer_date  \
1               2016-10-18 13:14:51                           NaT   
41              2016-10-09 17:06:52           2016-10-26 13:21:41   
72              2016-10-28 13:28:35           2016-11-08 14:03:58   
79              2016-10-09 03:45:42           2016-10-16 10:41:50   
80              2016-10-09 03:45:42           2016-10-16 10:41:50   
...                             ...                           ...   
102986          2018-08-27 13:25:00           2018-08-29 17:28:54   
102990          2018-08-27 14:15:00           2018-08-30 17:44:36   
103033          2018-08-27 15:03:00           2018-08-28 16:06:49   
103063          2018-08-29 15:10:00           2018-08-30 15:42:35   
103151          2018-08-29 13:00:00           2018-08-30 18:32:36   

       order_estimated_delivery_date  in_order_items  \
1                         2016-10-20            True   
41                        2016-11-28            True   
72                        2016-12-06            True   
79                        2016-11-28            True   
80                        2016-11-28            True   
...                              ...             ...   
102986                    2018-09-11            True   
102990                    2018-09-11            True   
103033                    2018-09-06            True   
103063                    2018-09-27            True   
103151                    2018-09-12            True   

                              product_id  order_item_id  \
1       f293394c72c9b5fafd7023301fc21fc2            2.0   
41      8f1e43e0e3374ed8c84df748034d6179            2.0   
72      dc82d7e8f99966cb4c641f7ea5590eaf            2.0   
79      66e08854f9319c6aa4874ca435598331            3.0   
80      d48e56be99464cbf285a9da17f618165            4.0   
...                                  ...            ...   
102986  e869c12a2f4763dbfc1a8398d20a1213            1.0   
102990  df4bcf4a328103951507064660c445e9            2.0   
103033  fcf311603fa276d415c006188e7bd070            1.0   
103063  b6b220a6c539cb65be34eed6fea8fe63            3.0   
103151  7001d71d1ad858e07e5a341649412e76            4.0   

                               seller_id shipping_limit_date  price

In [327]:
prep_orders_and_order_items['order_id'].duplicated().value_counts()

order_id
False    99441
True      3759
Name: count, dtype: int64

In [328]:
duplicate_ids = prep_orders_and_order_items[
    prep_orders_and_order_items['order_id'].duplicated(keep=False)
]['order_id'].unique()

duplicate_ids


array(['2e7a8482f6fb09756ca50c10d7bfc047',
       '51725d3e4bdfc97e28b40543310da8a3',
       '8b388d845a3e8700444b971fa2bef902', ...,
       'cb4d0ce750b3542055ae832b4d3bcca4',
       'a63d6f123bbccd71286ef76811cb2a3b',
       'd4fae577806d683110e00e18a5e181be'], shape=(3236,), dtype=object)

In [329]:
prep_orders_and_order_items[prep_orders_and_order_items['order_id']== 'ca3625898fbd48669d50701aba51cd5f']

order_id                       customer_id  \
99774  ca3625898fbd48669d50701aba51cd5f  0d861a5e4dd6a9079d89e1330848f0ab   
99775  ca3625898fbd48669d50701aba51cd5f  0d861a5e4dd6a9079d89e1330848f0ab   
99776  ca3625898fbd48669d50701aba51cd5f  0d861a5e4dd6a9079d89e1330848f0ab   
99777  ca3625898fbd48669d50701aba51cd5f  0d861a5e4dd6a9079d89e1330848f0ab   
99778  ca3625898fbd48669d50701aba51cd5f  0d861a5e4dd6a9079d89e1330848f0ab   
99779  ca3625898fbd48669d50701aba51cd5f  0d861a5e4dd6a9079d89e1330848f0ab   
99780  ca3625898fbd48669d50701aba51cd5f  0d861a5e4dd6a9079d89e1330848f0ab   
99781  ca3625898fbd48669d50701aba51cd5f  0d861a5e4dd6a9079d89e1330848f0ab   

      order_status order_purchase_timestamp   order_approved_at  \
99774    delivered      2018-08-12 02:11:20 2018-08-12 02:25:07   
99775    delivered      2018-08-12 02:11:20 2018-08-12 02:25:07   
99776    delivered      2018-08-12 02:11:20 2018-08-12 02:25:07   
99777    delivered      2018-08-12 02:11:20 2018-08-12 02:25:07   
99778    delivered      2018-08-12 02:11:20 2018-08-12 02:25:07   
99779    delivered      2018-08-12 02:11:20 2018-08-12 02:25:07   
99780    delivered      2018-08-12 02:11:20 2018-08-12 02:25:07   
99781    delivered      2018-08-12 02:11:20 2018-08-12 02:25:07   

      order_delivered_carrier_date order_delivered_customer_date  \
99774          2018-08-13 11:15:00           2018-08-21 12:18:48   
99775          2018-08-13 11:15:00           2018-08-21 12:18:48   
99776          2018-08-13 11:15:00           2018-08-21 12:18:48   
99777          2018-08-13 11:15:00           2018-08-21 12:18:48   
99778          2018-08-13 11:15:00           2018-08-21 12:18:48   
99779          2018-08-13 11:15:00           2018-08-21 12:18:48   
99780          2018-08-13 11:15:00           2018-08-21 12:18:48   
99781          2018-08-13 11:15:00           2018-08-21 12:18:48   

      order_estimated_delivery_date  in_order_items  \
99774                    2018-09-13            True   
99775                    2018-09-13            True   
99776                    2018-09-13            True   
99777                    2018-09-13            True   
99778                    2018-09-13            True   
99779                    2018-09-13            True   
99780                    2018-09-13            True   
99781                    2018-09-13            True   

                             product_id  order_item_id  \
99774  0cf2faf9749f53924cea652a09d8e327            2.0   
99775  0de59eddc63167215c972b0d785ffa7b            4.0   
99776  1065e0ebef073787a7bf691924c60eeb            1.0   
99777  21b524c4c060169fa75ccf08c7da4627            8.0   
99778  309dd69eb83cea38c51709d62befe1a4           10.0   
99779  4a5c3967bfd3629fe07ef4d0cc8c3818            6.0   
99780  5dae498eff2d80057f56122235a36aff            5.0   
99781  678c229b41c0e497d35a25a8be1cc631            7.0   

                              seller_id shipping_limit_date  price  \
99774  0b35c634521043bf4b47e21547b99ab5 2018-08-16 02:25:07   33.9   
99775  0b35c634521043bf4b47e21547b99ab5 2018-08-16 02:25:07  159.0   
99776  0b35c634521043bf4b47e21547b99ab5 2018-08-16 02:25:07  309.0   
99777  0b35c634521043bf4b47e21547b99ab5 2018-08-16 02:25:07   63.7   
99778  0b35c634521043bf4b47e21547b99ab5 2018-08-16 02:25:07   56.0   
99779  0b35c634521043bf4b47e21547b99ab5 2018-08-16 02:25:07  109.9   
99780  888faa8bfb0b159c37de6d898b961c31 2018-08-17 02:25:07   95.9   
99781  888faa8bfb0b159c37de6d898b961c31 2018-08-17 02:25:07   95.9   

       freight_value in_orders  
99774           1.84      True  
99775           3.67      True  
99776           1.84      True  
99777           0.15      True  
99778           3.68      True  
99779           0.15      True  
99780           0.15      True  
99781           0.15      True

In [330]:
prep_orders_and_order_items = prep_orders_and_order_items.drop (columns = [
    'product_id',
    'order_item_id',
    'price',
    'freight_value'
    ])
prep_orders_and_order_items.head()

order_id                       customer_id  \
0  2e7a8482f6fb09756ca50c10d7bfc047  08c5351a6aca1c1589a38f244edeee9d   
1  2e7a8482f6fb09756ca50c10d7bfc047  08c5351a6aca1c1589a38f244edeee9d   
2  e5fa5a7210941f7d56d0208e4e071d35  683c54fc24d40ee9f8a6fc179fd9856c   
3  809a282bbd5dbcabb6f2f724fca862ec  622e13439d6b5a0b486c435618b2679e   
4  bfbd0f9bdef84302105ad712db648a6c  86dc2ffce2dfff336de2f386a786e574   

  order_status order_purchase_timestamp   order_approved_at  \
0      shipped      2016-09-04 21:15:19 2016-10-07 13:18:03   
1      shipped      2016-09-04 21:15:19 2016-10-07 13:18:03   
2     canceled      2016-09-05 00:15:34 2016-10-07 13:17:15   
3     canceled      2016-09-13 15:24:19 2016-10-07 13:16:46   
4    delivered      2016-09-15 12:16:38 2016-09-15 12:16:38   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2016-10-18 13:14:51                           NaT   
1          2016-10-18 13:14:51                           NaT   
2                          NaT                           NaT   
3                          NaT                           NaT   
4          2016-11-07 17:11:53           2016-11-09 07:47:38   

  order_estimated_delivery_date  in_order_items  \
0                    2016-10-20            True   
1                    2016-10-20            True   
2                    2016-10-28            True   
3                    2016-09-30           False   
4                    2016-10-04            True   

                          seller_id shipping_limit_date in_orders  
0  1554a68530182680ad5c8b042c3ab563 2016-10-26 18:25:19      True  
1  1554a68530182680ad5c8b042c3ab563 2016-10-26 18:25:19      True  
2  a425f92c199eb576938df686728acd20 2016-09-19 00:15:34      True  
3                               NaN                 NaT       NaN  
4  ecccfa2bb93b34a3bf033cc5d1dcdc69 2016-09-19 23:11:33      True

In [331]:
new_order = [
     'order_id',
     'order_status',
     'order_purchase_timestamp',
     'order_approved_at',
     'shipping_limit_date',
     'order_delivered_carrier_date',
     'order_delivered_customer_date',
     'order_estimated_delivery_date',
     'customer_id',
     'seller_id'
 ]
prep_orders_and_order_items = prep_orders_and_order_items[new_order]
prep_orders_and_order_items.head()

order_id order_status order_purchase_timestamp  \
0  2e7a8482f6fb09756ca50c10d7bfc047      shipped      2016-09-04 21:15:19   
1  2e7a8482f6fb09756ca50c10d7bfc047      shipped      2016-09-04 21:15:19   
2  e5fa5a7210941f7d56d0208e4e071d35     canceled      2016-09-05 00:15:34   
3  809a282bbd5dbcabb6f2f724fca862ec     canceled      2016-09-13 15:24:19   
4  bfbd0f9bdef84302105ad712db648a6c    delivered      2016-09-15 12:16:38   

    order_approved_at shipping_limit_date order_delivered_carrier_date  \
0 2016-10-07 13:18:03 2016-10-26 18:25:19          2016-10-18 13:14:51   
1 2016-10-07 13:18:03 2016-10-26 18:25:19          2016-10-18 13:14:51   
2 2016-10-07 13:17:15 2016-09-19 00:15:34                          NaT   
3 2016-10-07 13:16:46                 NaT                          NaT   
4 2016-09-15 12:16:38 2016-09-19 23:11:33          2016-11-07 17:11:53   

  order_delivered_customer_date order_estimated_delivery_date  \
0                           NaT                    2016-10-20   
1                           NaT                    2016-10-20   
2                           NaT                    2016-10-28   
3                           NaT                    2016-09-30   
4           2016-11-09 07:47:38                    2016-10-04   

                        customer_id                         seller_id  
0  08c5351a6aca1c1589a38f244edeee9d  1554a68530182680ad5c8b042c3ab563  
1  08c5351a6aca1c1589a38f244edeee9d  1554a68530182680ad5c8b042c3ab563  
2  683c54fc24d40ee9f8a6fc179fd9856c  a425f92c199eb576938df686728acd20  
3  622e13439d6b5a0b486c435618b2679e                               NaN  
4  86dc2ffce2dfff336de2f386a786e574  ecccfa2bb93b34a3bf033cc5d1dcdc69

#### in the merged table of orders and order items, there are 775 missing sellers for 103200 rows in total

In [332]:
prep_orders_and_order_items['seller_id'].isnull().sum()

np.int64(775)

In [333]:
# null_seller_rows = prep_orders_and_order_items[
#     prep_orders_and_order_items['seller_id'].isnull()
# ]
# null_seller_rows

In [334]:
prep_orders_and_order_items_seller= pd.merge(prep_orders_and_order_items, prep_seller, how='left', on='seller_id')

prep_orders_and_order_items_seller.head()

order_id order_status order_purchase_timestamp  \
0  2e7a8482f6fb09756ca50c10d7bfc047      shipped      2016-09-04 21:15:19   
1  2e7a8482f6fb09756ca50c10d7bfc047      shipped      2016-09-04 21:15:19   
2  e5fa5a7210941f7d56d0208e4e071d35     canceled      2016-09-05 00:15:34   
3  809a282bbd5dbcabb6f2f724fca862ec     canceled      2016-09-13 15:24:19   
4  bfbd0f9bdef84302105ad712db648a6c    delivered      2016-09-15 12:16:38   

    order_approved_at shipping_limit_date order_delivered_carrier_date  \
0 2016-10-07 13:18:03 2016-10-26 18:25:19          2016-10-18 13:14:51   
1 2016-10-07 13:18:03 2016-10-26 18:25:19          2016-10-18 13:14:51   
2 2016-10-07 13:17:15 2016-09-19 00:15:34                          NaT   
3 2016-10-07 13:16:46                 NaT                          NaT   
4 2016-09-15 12:16:38 2016-09-19 23:11:33          2016-11-07 17:11:53   

  order_delivered_customer_date order_estimated_delivery_date  \
0                           NaT                    2016-10-20   
1                           NaT                    2016-10-20   
2                           NaT                    2016-10-28   
3                           NaT                    2016-09-30   
4           2016-11-09 07:47:38                    2016-10-04   

                        customer_id                         seller_id  ...  \
0  08c5351a6aca1c1589a38f244edeee9d  1554a68530182680ad5c8b042c3ab563  ...   
1  08c5351a6aca1c1589a38f244edeee9d  1554a68530182680ad5c8b042c3ab563  ...   
2  683c54fc24d40ee9f8a6fc179fd9856c  a425f92c199eb576938df686728acd20  ...   
3  622e13439d6b5a0b486c435618b2679e                               NaN  ...   
4  86dc2ffce2dfff336de2f386a786e574  ecccfa2bb93b34a3bf033cc5d1dcdc69  ...   

  seller_state won_date business_segment_old lead_type has_company has_gtin  \
0           MG      NaT                 None      None        True     True   
1           MG      NaT                 None      None        True     True   
2           PR      NaT                 None      None        True     True   
3          NaN      NaT                  NaN       NaN         NaN      NaN   
4           PR      NaT                 None      None        True     True   

  business_type business_segment business_segment_pred  \
0          None             None         Home & Living   
1          None             None         Home & Living   
2          None             None         Home & Living   
3           NaN              NaN                   NaN   
4          None             None       Beauty & Health   

  business_segment_incl_pred  
0              Home & Living  
1              Home & Living  
2              Home & Living  
3                        NaN  
4            Beauty & Health  

[5 rows x 22 columns]

In [335]:
prep_orders_and_order_items_seller.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103200 entries, 0 to 103199
Data columns (total 22 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       103200 non-null  object        
 1   order_status                   103200 non-null  object        
 2   order_purchase_timestamp       103200 non-null  datetime64[ns]
 3   order_approved_at              103040 non-null  datetime64[ns]
 4   shipping_limit_date            102425 non-null  datetime64[ns]
 5   order_delivered_carrier_date   101398 non-null  datetime64[ns]
 6   order_delivered_customer_date  100195 non-null  datetime64[ns]
 7   order_estimated_delivery_date  103200 non-null  datetime64[ns]
 8   customer_id                    103200 non-null  object        
 9   seller_id                      102425 non-null  object        
 10  seller_zip_code_prefix         102425 non-null  object        
 11  

In [336]:
prep_orders_and_order_items_seller.columns

Index(['order_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'shipping_limit_date',
       'order_delivered_carrier_date', 'order_delivered_customer_date',
       'order_estimated_delivery_date', 'customer_id', 'seller_id',
       'seller_zip_code_prefix', 'seller_city', 'seller_state', 'won_date',
       'business_segment_old', 'lead_type', 'has_company', 'has_gtin',
       'business_type', 'business_segment', 'business_segment_pred',
       'business_segment_incl_pred'],
      dtype='object')

In [337]:
prep_orders_and_order_items_seller=prep_orders_and_order_items_seller.drop (columns = [
    'seller_zip_code_prefix',
    'seller_city',
    'seller_state',
    'won_date',
    'lead_type',
    'has_company',
    'has_gtin',
    'business_type'
    ])
prep_orders_and_order_items_seller.head()

order_id order_status order_purchase_timestamp  \
0  2e7a8482f6fb09756ca50c10d7bfc047      shipped      2016-09-04 21:15:19   
1  2e7a8482f6fb09756ca50c10d7bfc047      shipped      2016-09-04 21:15:19   
2  e5fa5a7210941f7d56d0208e4e071d35     canceled      2016-09-05 00:15:34   
3  809a282bbd5dbcabb6f2f724fca862ec     canceled      2016-09-13 15:24:19   
4  bfbd0f9bdef84302105ad712db648a6c    delivered      2016-09-15 12:16:38   

    order_approved_at shipping_limit_date order_delivered_carrier_date  \
0 2016-10-07 13:18:03 2016-10-26 18:25:19          2016-10-18 13:14:51   
1 2016-10-07 13:18:03 2016-10-26 18:25:19          2016-10-18 13:14:51   
2 2016-10-07 13:17:15 2016-09-19 00:15:34                          NaT   
3 2016-10-07 13:16:46                 NaT                          NaT   
4 2016-09-15 12:16:38 2016-09-19 23:11:33          2016-11-07 17:11:53   

  order_delivered_customer_date order_estimated_delivery_date  \
0                           NaT                    2016-10-20   
1                           NaT                    2016-10-20   
2                           NaT                    2016-10-28   
3                           NaT                    2016-09-30   
4           2016-11-09 07:47:38                    2016-10-04   

                        customer_id                         seller_id  \
0  08c5351a6aca1c1589a38f244edeee9d  1554a68530182680ad5c8b042c3ab563   
1  08c5351a6aca1c1589a38f244edeee9d  1554a68530182680ad5c8b042c3ab563   
2  683c54fc24d40ee9f8a6fc179fd9856c  a425f92c199eb576938df686728acd20   
3  622e13439d6b5a0b486c435618b2679e                               NaN   
4  86dc2ffce2dfff336de2f386a786e574  ecccfa2bb93b34a3bf033cc5d1dcdc69   

  business_segment_old business_segment business_segment_pred  \
0                 None             None         Home & Living   
1                 None             None         Home & Living   
2                 None             None         Home & Living   
3                  NaN              NaN                   NaN   
4                 None             None       Beauty & Health   

  business_segment_incl_pred  
0              Home & Living  
1              Home & Living  
2              Home & Living  
3                        NaN  
4            Beauty & Health

In [338]:
prep_orders_and_order_items_seller.to_csv('prep_orders_and_order_items_seller.csv', index=False)

### Seller scorecard: metric (handling time benchmarking)

In [339]:
# prep_orders_and_order_items_seller['handling_time']=prep_orders_and_order_items_seller['order_delivered_carrier_date']-prep_orders_and_order_items_seller['order_approved_at']
# prep_orders_and_order_items_seller.head()

In [340]:
# prep_orders_and_order_items_seller['handling_time'].isnull().sum()

In [341]:
# null_count = prep_orders_and_order_items_seller['handling_time'].isnull().sum()
# total_count = len(prep_orders_and_order_items_seller)

# percentage_null = (null_count / total_count) * 100
# percentage_null


In [342]:
# prep_orders_and_order_items_seller[prep_orders_and_order_items_seller['handling_time'].isnull()]

In [343]:
# null_ht = prep_orders_and_order_items_seller[prep_orders_and_order_items_seller['handling_time'].isnull()]
# total_null_ht = len(null_ht)
# total_null_ht

In [344]:
# null_ht_approved_count = null_ht['order_approved_at'].notnull().sum()
# null_ht_approved_count

In [345]:
# null_ht_not_approved_count = null_ht['order_approved_at'].isnull().sum()
# null_ht_not_approved_count

In [346]:
# # Percentage of null handling_time orders that are approved
# percentage_approved_null_ht = (null_ht_approved_count / total_null_ht) * 100
# percentage_approved_null_ht

In [347]:
# # Percentage of null handling_time orders that are NOT approved
# percentage_not_approved_null_ht = (null_ht_not_approved_count / total_null_ht) * 100
# percentage_not_approved_null_ht

In [348]:
# approved = prep_orders[prep_orders['order_approved_at'].notnull()]
# approved

In [349]:
# not_approved = prep_orders[prep_orders['order_approved_at'].isnull()]
# not_approved

In [350]:
# prep_orders_and_order_items_seller = prep_orders_and_order_items_seller.dropna(
#     subset=['handling_time']
# )

### Seller scorecard: metric (On-time shipping rate benchmarking & Avg shipping delay benchmarking)

In [351]:
# prep_orders_and_order_items_seller.head()

In [352]:
#prep_orders_and_order_items_seller['delivered_on_time'] = prep_orders_and_order_items_seller['order_delivered_customer_date'] <= prep_orders_and_order_items_seller['order_estimated_delivery_date']

#### out of 103200 rows there are 3005 missing data of order_delivered_customer_date, which makes 2.92% of the whole orders.

In [353]:
# null_delivered = prep_orders_and_order_items_seller['order_delivered_customer_date'].isnull().sum()
# null_delivered

In [354]:
# total_rows = len(prep_orders_and_order_items_seller)
# total_rows

In [355]:
# percentage_null = (null_delivered / total_rows) * 100
# print(f"Percentage of undelivered orders: {percentage_null:.2f}%")

### Seller scorecard: metric (Average delivery delay benchmarking)

In [356]:
# prep_orders_and_order_items_seller['delay_days'] = (
#     prep_orders_and_order_items_seller['order_delivered_customer_date'] -
#     prep_orders_and_order_items_seller['order_estimated_delivery_date']
# ).dt.days
# prep_orders_and_order_items_seller.head()

In [357]:
# prep_orders_and_order_items_seller.tail()

In [358]:
# def classify_delivery(row):
#     if pd.isna(row['order_delivered_customer_date']):
#         return 'Undelivered'
#     elif row['delay_days'] <= 0:
#         return 'On Time'
#     elif 1 <= row['delay_days'] <= 3:
#         return 'Small Delay'
#     else:
#         return 'Big Delay'

# prep_orders_and_order_items_seller['delivery_status'] = \
#     prep_orders_and_order_items_seller.apply(classify_delivery, axis=1)


In [359]:
# prep_orders_and_order_items_seller

In [360]:
# prep_orders_and_order_items_seller['delivery_status'].value_counts()

#### exclude undelivered orders when calculating the on time delivery rate

In [361]:
# delivered_df = prep_orders_and_order_items_seller[
#     prep_orders_and_order_items_seller['order_delivered_customer_date'].notnull()
# ]

In [362]:
# undelivered_df = prep_orders_and_order_items_seller[
#     prep_orders_and_order_items_seller['order_delivered_customer_date'].isnull()
# ]

#### Benchmarking: Avg on time delivery rate

In [363]:
# otdr_of_all_sellers = delivered_df['delivered_on_time'].mean() * 100
# print(f"On-Time Delivery Rate: {otdr:.2f}%")

In [364]:
# otdr_by_seller = (
#     delivered_df
#     .groupby('seller_id')
#     .agg(
#         total_deliveries=('delivered_on_time', 'count'),
#         on_time_deliveries=('delivered_on_time', 'sum'),
#         otdr_percent=('delivered_on_time', lambda x: x.mean() * 100)
#     )
#     .reset_index()
# )

### Seller scorecard: metric (Shipping Limit Compliance Rate (SLC Rate))


In [365]:
# prep_orders_and_order_items_seller['shipped_before_limit'] = (
#     prep_orders_and_order_items_seller['order_delivered_carrier_date']
#     <= prep_orders_and_order_items_seller['shipping_limit_date']
# )
# prep_orders_and_order_items_seller.head()

In [366]:
prep_orders_and_order_items_seller.to_csv('mart_orders_and_order_items_seller.csv', index=False)

In [367]:
prep_orders_and_order_items_seller

order_id order_status  \
0       2e7a8482f6fb09756ca50c10d7bfc047      shipped   
1       2e7a8482f6fb09756ca50c10d7bfc047      shipped   
2       e5fa5a7210941f7d56d0208e4e071d35     canceled   
3       809a282bbd5dbcabb6f2f724fca862ec     canceled   
4       bfbd0f9bdef84302105ad712db648a6c    delivered   
...                                  ...          ...   
103195  392ed9afd714e3c74767d0c4d3e3f477     canceled   
103196  616fa7d4871b87832197b2a137a115d2     canceled   
103197  a2ac6dad85cf8af5b0afb510a240fe8c     canceled   
103198  b059ee4de278302d550a3035c4cdb740     canceled   
103199  10a045cdf6a5650c21e9cfeb60384c16     canceled   

       order_purchase_timestamp   order_approved_at shipping_limit_date  \
0           2016-09-04 21:15:19 2016-10-07 13:18:03 2016-10-26 18:25:19   
1           2016-09-04 21:15:19 2016-10-07 13:18:03 2016-10-26 18:25:19   
2           2016-09-05 00:15:34 2016-10-07 13:17:15 2016-09-19 00:15:34   
3           2016-09-13 15:24:19 2016-10-07 13:16:46                 NaT   
4           2016-09-15 12:16:38 2016-09-15 12:16:38 2016-09-19 23:11:33   
...                         ...                 ...                 ...   
103195      2018-09-29 09:13:03                 NaT                 NaT   
103196      2018-10-01 15:30:09                 NaT                 NaT   
103197      2018-10-03 18:55:29                 NaT                 NaT   
103198      2018-10-16 20:16:02                 NaT                 NaT   
103199      2018-10-17 17:30:18                 NaT                 NaT   

       order_delivered_carrier_date order_delivered_customer_date  \
0               2016-10-18 13:14:51                           NaT   
1               2016-10-18 13:14:51                           NaT   
2                               NaT                           NaT   
3                               NaT                           NaT   
4               2016-11-07 17:11:53           2016-11-09 07:47:38   
...                             ...                           ...   
103195                          NaT                           NaT   
103196                          NaT                           NaT   
103197                          NaT                           NaT   
103198                          NaT                           NaT   
103199                          NaT                           NaT   

       order_estimated_delivery_date                       customer_id  \
0                         2016-10-20  08c5351a6aca1c1589a38f244edeee9d   
1                         2016-10-20  08c5351a6aca1c1589a38f244edeee9d   
2                         2016-10-28  683c54fc24d40ee9f8a6fc179fd9856c   
3                         2016-09-30  622e13439d6b5a0b486c435618b2679e   
4                         2016-10-04  86dc2ffce2dfff336de2f386a786e574   
...                              ...                               ...   
103195                    2018-10-15  2823ffda607a2316375088e0d00005ec   
103196                    2018-10-23  bf6181a85bbb4115736c0a8db1a53be3   
103197                    2018-10-16  4c2ec60c29d10c34bd49cb88aa85cfc4   
103198                    2018-11-12  856336203359aa6a61bf3826f7d84c49   
103199                    2018-10-30  a4b417188addbc05b26b72d5e44837a1   

                               seller_id business_segment_old  \
0       1554a68530182680ad5c8b042c3ab563                 None   
1       1554a68530182680ad5c8b042c3ab563                 None   
2       a425f92c199eb576938df686728acd20                 None   
3                                    NaN                  NaN   
4       ecccfa2bb93b34a3bf033cc5d1dcdc69                 None   
...                                  ...                  ...   
103195                               NaN                  NaN   
103196                               NaN                  NaN   
103197                               NaN                  NaN   
103198                               NaN                  NaN   
103199       

In [368]:
prep_orders_and_order_items_seller.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103200 entries, 0 to 103199
Data columns (total 14 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       103200 non-null  object        
 1   order_status                   103200 non-null  object        
 2   order_purchase_timestamp       103200 non-null  datetime64[ns]
 3   order_approved_at              103040 non-null  datetime64[ns]
 4   shipping_limit_date            102425 non-null  datetime64[ns]
 5   order_delivered_carrier_date   101398 non-null  datetime64[ns]
 6   order_delivered_customer_date  100195 non-null  datetime64[ns]
 7   order_estimated_delivery_date  103200 non-null  datetime64[ns]
 8   customer_id                    103200 non-null  object        
 9   seller_id                      102425 non-null  object        
 10  business_segment_old           4615 non-null    object        
 11  

In [369]:
# Let's load values from the .env file
from dotenv import dotenv_values

config = dotenv_values()

# define variables for the login
pg_user = config['POSTGRES_USER']  # align the key label with your .env file !
pg_host = config['POSTGRES_HOST']
pg_port = config['POSTGRES_PORT']
pg_db = config['POSTGRES_DB']
pg_schema = config['POSTGRES_SCHEMA']
pg_pass = config['POSTGRES_PASS']

In [370]:
# updating the url
url = f'postgresql://{pg_user}:{pg_pass}@{pg_host}:{pg_port}/{pg_db}' #the same like version 1

# let's switch the logging off again. .
engine = create_engine(url, echo=False) #the same like version 1


# writing dataframe to DB : Pandas Dataframe to DB Table in my own Schema
prep_orders_and_order_items_seller.to_sql(name = 'mart_seller_fulfillment_scorecard', 
                       con = engine, 
                       schema = pg_schema, # pandas is allowing to specify, in which schema the table shall be created
                       if_exists='replace', 
                       index=False
                      )

200